### PR TITLE
Fix error in 'qml.data.load()` when using 'full' parameter value

### LINKED ISSUE #4270: [BUG] `qml.data.load()` fails when using 'full' parameter value
### Expected behavior

`qml.data.load('qspin',sysname='Heisenberg',periodicity='open',lattice='full',layout='2x2')`
Should return the Heisenberg dataset corresponding to the `open` periodicity, `rectangular` lattice, and `2x2` layout.

### Actual behavior

`qml.data.load('qspin',sysname='Heisenberg',periodicity='open',lattice='full',layout='2x2')`

Raises `ValueError: layout value of '2x2' is not available. Available values are ['1x16', '1x4', '1x8']`

### Additional information

_No response_

### Source code

```shell
qml.data.load('qspin',sysname='Heisenberg',periodicity='open',lattice='full',layout='1x4') # works

qml.data.load('qspin',sysname='Heisenberg',periodicity='open',lattice='full',layout='2x2') # fails
```

### Tracebacks

_No response_

### System information

```shell
Name: PennyLane
Version: 0.31.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: .venv/lib/python3.8/site-packages
Editable project location: 
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml
Required-by: PennyLane-Lightning

Platform info:           Linux-5.15.0-73-generic-x86_64-with-glibc2.29
Python version:          3.8.10
Numpy version:           1.23.5
Scipy version:           1.10.1
Installed devices:
- default.gaussian (PennyLane-0.31.0.dev0)
- default.mixed (PennyLane-0.31.0.dev0)
- default.qubit (PennyLane-0.31.0.dev0)
- default.qubit.autograd (PennyLane-0.31.0.dev0)
- default.qubit.jax (PennyLane-0.31.0.dev0)
- default.qubit.tf (PennyLane-0.31.0.dev0)
- default.qubit.torch (PennyLane-0.31.0.dev0)
- default.qutrit (PennyLane-0.31.0.dev0)
- null.qubit (PennyLane-0.31.0.dev0)
- lightning.qubit (PennyLane-Lightning-0.30.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by brownj85**
**Context:**
`qml.data.load()` did not properly handle specified parameters that come after `'full'`.  For example, the following
call should return all qspin datasets matching `sysname='Heisenberg', periodicity='open', layout='2x2'`, for any value
of `lattice`:

```
qml.data.load('qspin',sysname='Heisenberg',periodicity='open',lattice='full',layout='2x2')
```
Instead, we get an exception, even though matching datasets do exist:
```
ValueError: layout value of '2x2' is not available. Available values are ['1x16', '1x4', '1x8']
```

This happens because the `find()` function fails as soon as it can't find a matching dataset for any of the parameters. When it checks the `'chain'` lattice, it can't find `'2x2'` and throws an exception, even though `'2x2'` is available for `'rectangular`'.

**Description of the Change:**
`Foldermap.find()` will only throw an exception if it cannot find any matching datasets for a given parameter. This does mean that the user can provide invalid parameters, as long as one of them matches. For example:

```
qml.data.load('qspin',sysname='Heisenberg',periodicity='open',lattice='full',layout=['2x2', '17x32'])
```
Will not result in an error.

**Benefits:**
- `qml.data.load()` can handle full parameters that come before specific parameters.

**Possible Drawbacks:**
- The user will not get immediate feedback for erroneous parameters
- Users are no logner guaranteed to get a dataset for every combination of (fully qualified) parameters. For example, `qml.data.load('qchem', molname='H2', basis=['STO-3G', 'CC-PVDZ'], bondlength=['0.5', '0.7'])` will return 4 datasets. But if one those is removed, 3 will be returned without throwing an error.

**Related GitHub Issues:**
Fixes #4270

### CODE REVIEW

File: tests/data/data_manager/test_foldermap.py [hunk 1]
```diff
@@ -125,6 +127,38 @@ class TestFolderMapView:
...
+                    ),
+                ],
+            ),
+            (
+                {
+                    "missing_default": DEFAULT,
+                    "molname": "O2",
+                    "basis": FULL,
+                    "bondlength": ["0.6", "200"],
+                },
```
Comment (timmysilv) [inline tests/data/data_manager/test_foldermap.py:145]: is this the test-case that fails without the change/was it failing locally without the fix?

**Review by timmysilv — COMMENTED** [review]
looks good! just clarifying something you said:
> ...`basis=['STO-3G', 'CC-PVDZ'], bondlength=['0.5', '0.7'])` will return 4 datasets. But if one those is removed, 3 will be returned without throwing an error. 

if one what is removed? and which 3 will be returned? I'm not entirely sure what the expected behaviour is there.

**Comment by brownj85** [comment]
> if one what is removed? and which 3 will be returned? I'm not entirely sure what the expected behaviour is there.

So before this PR, if `H2_CC-PVDZ_0.7` was removed from the datasets bucket (and the foldermap), calling
`qml.data.load('qchem', molname='H2', basis=['STO-3G', 'CC-PVDZ'], bondlength=['0.5', '0.7'])` would cause an exception, because one of the requested parameter combinations doesn't exist. After this PR it'll return the other 3 datasets without an exception. An exception is only raised if **none** of the requested parameter combinations exist.

There's not really an 'expected' behaviour here (afaik) so we're not breaking any promises.

File: tests/data/data_manager/test_foldermap.py (hunk 1 above)
Comment (brownj85) [reply tests/data/data_manager/test_foldermap.py:145]: The one just above it actually - the [ticket](https://app.shortcut.com/xanaduai/story/40483/bug-qml-data-load-fails-when-using-full-parameter-value) explains it.

**Review by timmysilv — APPROVED** [review]
ahh gotcha - you mean removed from the actual database, makes sense! looks great :shipit:

**Review by DSGuala — APPROVED** [review]
Looks good to me, not getting the error anymore :100:

**Pull request merged**